xbmc/pvr/providers/PVRProvider.cpp

```diff
@@ -0,0 +1,363 @@
+  {
+    m_strProviderName = strProviderName;
+    return true;
+  }
+
+  return false;
+}
+
+PVR_PROVIDER_TYPE CPVRProvider::GetProviderType() const
+{
+  CSingleLock lock(m_critSection);
+  return m_providerType;
+}
+
+bool CPVRProvider::SetProviderType(PVR_PROVIDER_TYPE providerType)
+{
+  CSingleLock lock(m_critSection);
+  if (m_providerType != providerType)
+  {
+    m_providerType = providerType;
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRProvider::GetIconPath() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strIconPath;
+}
+
+bool CPVRProvider::SetIconPath(const std::string& strIconPath)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strIconPath != strIconPath)
+  {
+    m_strIconPath = strIconPath;
+    return true;
+  }
+
+  return false;
+}
+
+namespace
+{
+
+const std::vector<std::string> Tokenize(const std::string& str)
+{
+  return StringUtils::Split(str.c_str(), PROVIDER_STRING_TOKEN_SEPARATOR);
+}
+
+const std::string DeTokenize(const std::vector<std::string>& tokens)
+{
+  return StringUtils::Join(tokens, PROVIDER_STRING_TOKEN_SEPARATOR);
+}
+
+} // unamed namespace
+
+std::vector<std::string> CPVRProvider::GetCountries() const
+{
+  CSingleLock lock(m_critSection);
+
+  return Tokenize(m_strCountries);
+}
+
+bool CPVRProvider::SetCountries(const std::vector<std::string>& countries)
+{
+  CSingleLock lock(m_critSection);
+  const std::string strCountries = DeTokenize(countries);
+  if (m_strCountries != strCountries)
+  {
+    m_strCountries = strCountries;
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRProvider::GetCountriesDBString() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strCountries;
+}
+
+bool CPVRProvider::SetCountriesFromDBString(const std::string& strCountries)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strCountries != strCountries)
+  {
+    m_strCountries = strCountries;
+    return true;
+  }
+
+  return false;
+}
+
+std::vector<std::string> CPVRProvider::GetLanguages() const
+{
+  CSingleLock lock(m_critSection);
+  return Tokenize(m_strLanguages);
+}
+
+bool CPVRProvider::SetLanguages(const std::vector<std::string>& languages)
+{
+  CSingleLock lock(m_critSection);
+  const std::string strLanguages = DeTokenize(languages);
+  if (m_strLanguages != strLanguages)
+  {
+    m_strLanguages = strLanguages;
+    return true;
+  }
+
+  return false;
+}
+
+std::string CPVRProvider::GetLanguagesDBString() const
+{
+  CSingleLock lock(m_critSection);
+  return m_strLanguages;
+}
+
+bool CPVRProvider::SetLanguagesFromDBString(const std::string& strLanguages)
+{
+  CSingleLock lock(m_critSection);
+  if (m_strLanguages != strLanguages)
+  {
+    m_strLanguages = strLanguages;
+    return true;
+  }
+
+  return false;
+}
+
+bool CPVRProvider::Persist(bool updateRecord /* = false */)
+{
+  CSingleLock lock(m_critSection);
+  const std::shared_ptr<CPVRDatabase> database = CServiceBroker::GetPVRManager().GetTVDatabase();
```

Comment:
I would obtain the lock as late as possible, means after getting `database`.

Reply:
Done